### PR TITLE
Migrate `ReferenceField` fields from setup types to `UIDReferenceField`

### PR DESCRIPTION
**Pull request by xispa**
## Description of the issue/feature this PR addresses

This Pull Request migrates the `ReferenceField` fields to `UIDReferenceField` from objects other than `AnalysisRequest` (that were migrated thanks to https://github.com/senaite/senaite.core/pull/2209)

## Current behavior before PR

Objects that use AT's `ReferenceField` are still present

## Desired behavior after PR is merged

No AT's `ReferenceField` remain

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

### CODE REVIEW

File: src/bika/lims/content/referencesample.py [hunk 1]
```diff
@@ -42,10 +43,9 @@
 from zope.interface import implements
 
 schema = BikaSchema.copy() + Schema((
-    ReferenceField('ReferenceDefinition',
+    UIDReferenceField('ReferenceDefinition',
         schemata = 'Description',
         allowed_types = ('ReferenceDefinition',),
-        relationship = 'ReferenceSampleReferenceDefinition',
         referenceClass = HoldingReference,
```
Comment (ramonski) [inline src/bika/lims/content/referencesample.py]: I think the `referenceClass` can be removed

File: src/bika/lims/content/referencesample.py [hunk 2]
```diff
@@ -69,10 +69,9 @@
             description=_("Samples of this type should be treated as hazardous"),
         ),
     ),
-    ReferenceField('Manufacturer',
+    UIDReferenceField('Manufacturer',
         schemata = 'Description',
         allowed_types = ('Manufacturer',),
-        relationship = 'ReferenceSampleManufacturer',
         vocabulary = "getManufacturers",
         referenceClass = HoldingReference,
```
Comment (ramonski) [inline src/bika/lims/content/referencesample.py]: here as well

File: src/bika/lims/content/report.py [hunk 3]
```diff
@@ -41,9 +42,8 @@
             description=_("Report type"),
         ),
     ),
-    ReferenceField('Client',
+    UIDReferenceField('Client',
         allowed_types = ('Client',),
-        relationship = 'ReportClient',
         widget = ReferenceWidget(
```
Comment (ramonski) [inline src/bika/lims/content/report.py]: I think we need to take care here. Is the default `ReferenceWidget` coming from `Products.Archetypes` compatible with our `UIDReferenceField`?

**Review by ramonski — CHANGES_REQUESTED** [review]
In my opinion we could maintain backreferences with their relationships for setup objects, because there should not be so many and I think it might be worth.

As a rule of thumb, I would remove backrefs if the object is contained in the referenced object, e.g. like with sample -> client.

I'm just not a big fan of this half/half solutions with no clear rules behind it...

**Comment by xispa** [comment]
> In my opinion we could maintain backreferences with their relationships for setup objects, because there should not be so many and I think it might be worth.
> 
> 
> I'm just not a big fan of this half/half solutions with no clear rules behind it...

I disagree. I think the rule should be to not keep back-references unless strictly necessary. And since the legacy `ReferenceField` from AT does not come with this behavior by default, I think is way better to rely on a single strategy and apply same approach to all objects.


> As a rule of thumb, I would remove backrefs if the object is contained in the referenced object, e.g. like with sample -> client.

With such rule, the approved behavior of not keeping back-references by default would become a rarity, only for `AnalysisRequest` content type. And still, note that with such rule, we should keep backreferences from e.g. `SampleType` objects to `AnalysisRequest`, `AnalysisProfile` to `AnalysisRequest`, etc...... This means that, besides ending up with setup objects with Persistent lists made of thousands of UIDs, each time an `AnalysisRequest` is created, plenty of objects are updated to simply add a UID into those lists. Thus, the chances of conflicts increases, as well as the time it takes fhe creation of an `AnalysisRequest` object.

File: src/bika/lims/content/report.py (hunk 3 above)
Comment (xispa) [reply src/bika/lims/content/report.py]: This `ReferenceWidget` is never rendered, the value for this field is always set programmatically here: https://github.com/senaite/senaite.core/blob/ea8189a638d20daedbbf6bea21f37851bf42b174/src/bika/lims/browser/reports/__init__.py#L301

However, will update the import to our `ReferenceWidget` in accordance (and make it `visible=False`)

File: src/bika/lims/content/referencesample.py (hunk 1 above)
Comment (xispa) [reply src/bika/lims/content/referencesample.py]: Yep, will do so

File: src/bika/lims/content/report.py (hunk 3 above)
Comment (xispa) [reply src/bika/lims/content/report.py]: Done with https://github.com/senaite/senaite.core/pull/2221/commits/dc4f78ab2eed5d78b0febba038730a3a22ba3457

File: src/bika/lims/content/referencesample.py (hunk 1 above)
Comment (xispa) [reply src/bika/lims/content/referencesample.py]: I've removed all `referenceClass` properties that do no longer make sense, as well as `vocabulary_display_path_bound` and `checkbox_bound` properties

**Review by ramonski — APPROVED** [review]

**Pull request merged**